Pod/Classes/WPPHAssetDataSource.m

```diff
@@ -450,40 +450,86 @@ - (void)addAssetWithChangeRequest:(PHAssetChangeRequest *(^)(void))changeRequest
 {
     NSParameterAssert(changeRequestBlock);
     __block NSString * assetIdentifier = nil;
-    [[PHPhotoLibrary sharedPhotoLibrary] performChanges:^{
-        // Request creating an asset from the image.
-        PHAssetChangeRequest *createAssetRequest = changeRequestBlock();
-        PHObjectPlaceholder *assetPlaceholder = [createAssetRequest placeholderForCreatedAsset];
-        assetIdentifier = [assetPlaceholder localIdentifier];
-        if ([self.activeAssetsCollection canPerformEditOperation:PHCollectionEditOperationAddContent]) {
-            // Request editing the album.
-            PHAssetCollectionChangeRequest *albumChangeRequest = [PHAssetCollectionChangeRequest changeRequestForAssetCollection:self.activeAssetsCollection];
-            [albumChangeRequest addAssets:@[ assetPlaceholder ]];
-        }
-    } completionHandler:^(BOOL success, NSError *error) {
-        if (!success) {
-            if (completionBlock){
-                dispatch_async(dispatch_get_main_queue(), ^{
-                    completionBlock(nil, error);
-                });
+
+    [self checkPermissionStatus:^(PHAuthorizationStatus status) {
+
+        /// Starting from iOS 15.2 we should do the registration
```

Comment:
There's a lot of similarity here between this implementation in `addAssetWithChangeRequest:` and `loadDataWithOptions:success:failure:`. Could we do anything to consolidate the two or share code at all? We probably also don't need this exact comment duplicated.

I wonder if we could move the call to `registerChangeObserver` into `checkPermissionStatus` and call it before we call the handler block, seeing as we now need it in two places?

Reply:
I've created a function for handling `registerChangeObserver` + will drop you a message to discuss the other(s). Thanks!